// @cognima/banners/src/constants.js
"use strict";

/**
 * Módulo de Constantes para @cognima/banners
 * 
 * Este módulo contém constantes compartilhadas entre os diferentes tipos de banners,
 * incluindo cores padrão, configurações de layout e outros valores reutilizáveis.
 * 
 * @author Cognima Team (melhorado)
 * @version 2.0.0
 */

// --- Cores Padrão ---
const DEFAULT_COLORS = {
    // Cores de plataformas
    discord: {
        primary: "#5865F2",
        secondary: "#2F3136",
        dark: "#36393F",
        light: "#FFFFFF",
        muted: "#B9BBBE",
        success: "#57F287",
        danger: "#ED4245",
        warning: "#FEE75C"
    },
    whatsapp: {
        primary: "#128C7E",
        secondary: "#25D366",
        dark: "#075E54",
        light: "#FFFFFF",
        muted: "#ECECEC"
    },
    instagram: {
        primary: "#E1306C",
        secondary: "#F56040",
        tertiary: "#FCAF45",
        dark: "#262626",
        light: "#FFFFFF",
        muted: "#8E8E8E",
        link: "#00376B"
    },
    youtube: {
        primary: "#FF0000",
        secondary: "#282828",
        dark: "#0F0F0F",
        light: "#FFFFFF",
        muted: "#AAAAAA"
    },
    twitter: {
        primary: "#1DA1F2",
        secondary: "#14171A",
        dark: "#657786",
        light: "#FFFFFF",
        muted: "#AAB8C2"
    },
    telegram: {
        primary: "#0088CC",
        secondary: "#32AFED",
        dark: "#1E1E1E",
        light: "#FFFFFF",
        muted: "#D2D9DF"
    },
    
    // Cores genéricas
    background: {
        dark: "#1A1A1A",
        light: "#FFFFFF",
        gray: "#F5F5F5"
    },
    text: {
        dark: "#333333",
        light: "#FFFFFF",
        muted: "#888888"
    },
    accent: {
        blue: "#3498DB",
        green: "#2ECC71",
        red: "#E74C3C",
        purple: "#9B59B6",
        orange: "#F39C12",
        teal: "#1ABC9C"
    },
    gradient: {
        blue: {
            start: "#3498DB",
            end: "#2980B9"
        },
        purple: {
            start: "#9B59B6",
            end: "#8E44AD"
        },
        green: {
            start: "#2ECC71",
            end: "#27AE60"
        },
        orange: {
            start: "#F39C12",
            end: "#D35400"
        },
        red: {
            start: "#E74C3C",
            end: "#C0392B"
        },
        dark: {
            start: "#34495E",
            end: "#2C3E50"
        }
    }
};

// --- Status de Usuário ---
const USER_STATUS = {
    online: {
        color: "#43B581",
        label: "Online"
    },
    idle: {
        color: "#FAA61A",
        label: "Ausente"
    },
    dnd: {
        color: "#F04747",
        label: "Não Perturbe"
    },
    streaming: {
        color: "#593695",
        label: "Transmitindo"
    },
    offline: {
        color: "#747F8D",
        label: "Offline"
    }
};

// --- Configurações de Layout ---
const LAYOUT = {
    cornerRadius: {
        none: 0,
        small: 5,
        medium: 10,
        large: 15,
        xlarge: 20,
        round: 9999
    },
    padding: {
        small: 10,
        medium: 20,
        large: 30
    },
    shadow: {
        light: "rgba(0, 0, 0, 0.1) 0px 1px 3px 0px",
        medium: "rgba(0, 0, 0, 0.2) 0px 3px 7px -1px",
        strong: "rgba(0, 0, 0, 0.3) 0px 10px 20px -2px"
    }
};

// --- Dimensões Padrão ---
const DEFAULT_DIMENSIONS = {
    welcome: {
        width: 700,
        height: 350
    },
    profile: {
        width: 600,
        height: 800
    },
    card: {
        width: 400,
        height: 200
    },
    banner: {
        width: 1200,
        height: 400
    },
    post: {
        width: 600,
        height: 600
    },
    thumbnail: {
        width: 1280,
        height: 720
    },
    story: {
        width: 720,
        height: 1280
    }
};

// --- Fontes Padrão ---
const FONT_SIZES = {
    xs: 12,
    sm: 14,
    md: 16,
    lg: 20,
    xl: 24,
    xxl: 32,
    xxxl: 48
};

// --- Exportações ---
module.exports = {
    DEFAULT_COLORS,
    USER_STATUS,
    LAYOUT,
    DEFAULT_DIMENSIONS,
    FONT_SIZES
};
